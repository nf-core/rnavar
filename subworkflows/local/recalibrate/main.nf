/*
========================================================================================
    RECALIBRATE
========================================================================================
*/

include { GATK4_APPLYBQSR as APPLYBQSR } from '../../../modules/nf-core/gatk4/applybqsr/main'
include { SAMTOOLS_INDEX               } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_STATS               } from '../../../modules/nf-core/samtools/stats/main'

workflow RECALIBRATE {
    take:
        skip_samtools  // boolean: true/false
        bam         // channel: [mandatory] bam
        dict        // channel: [mandatory] dict
        fai         // channel: [mandatory] fai
        fasta       // channel: [mandatory] fasta

    main:

    def ch_versions = Channel.empty()

    APPLYBQSR (
        bam,
        fasta,
        fai,
        dict
    )
    def bam_recalibrated = APPLYBQSR.out.bam
    ch_versions = ch_versions.mix(APPLYBQSR.out.versions.first())

    SAMTOOLS_INDEX(bam_recalibrated)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    def bam_indices = SAMTOOLS_INDEX.out.bai
        .mix(SAMTOOLS_INDEX.out.csi)
        .mix(SAMTOOLS_INDEX.out.crai)

    def bam_recalibrated_index = bam_recalibrated
        .join(bam_indices, failOnMismatch: true, failOnDuplicate: true)

    def bam_reports = Channel.empty()

    if (!skip_samtools) {
        SAMTOOLS_STATS(bam_recalibrated_index, [[], []])
        bam_reports = SAMTOOLS_STATS.out.stats
        ch_versions = ch_versions.mix(SAMTOOLS_STATS.out.versions.first())
    }


    emit:
    bam         = bam_recalibrated_index
    qc          = bam_reports

    versions    = ch_versions

}
