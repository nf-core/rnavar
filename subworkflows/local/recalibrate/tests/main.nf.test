nextflow_workflow {

    name "Test Subworkflow RECALIBRATE"
    script "../main.nf"
    workflow "RECALIBRATE"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/recalibrate"
    tag "samtools/stats"
    tag "samtools/index"
    tag "gatk4/applybqsr"

    setup {
        run("GATK4_BASERECALIBRATOR") {
            script "../../../../modules/nf-core/gatk4/baserecalibrator/main.nf"

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam.bai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.filtered_intervals.interval_list', checkIfExists: true)
                ])
                input[1] = Channel.value([
                    [id:"fasta"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = Channel.value([
                    [id:"fai"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = Channel.value([
                    [id:"dict"],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                ])
                input[4] = Channel.value([
                    [id:"known_sites"],
                    [
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz', checkIfExists: true)
                    ]
                ])
                input[5] = Channel.value([
                    [id:"known_sites"],
                    [
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz.tbi', checkIfExists: true)
                    ]
                ])
                """
            }
        }
    }

    test("bam - stats") {
        when {
            workflow {
                """
                input[0] = false
                input[1] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                .join(GATK4_BASERECALIBRATOR.out.table)
                .map { meta, bam, bai, table ->
                    [
                        meta,
                        bam,
                        bai,
                        table,
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.filtered_intervals.interval_list', checkIfExists: true)
                    ]
                }
                input[2] = Channel.value(
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                )
                input[3] = Channel.value(
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                )
                input[4] = Channel.value(
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        }
                        if(key == "bam") {
                            return [key, val.collect { [it[0], bam(it[1]).readsMD5, file(it[2].toString()).name]}]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }

    test("bam - no stats") {
        when {
            workflow {
                """
                input[0] = true
                input[1] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                .join(GATK4_BASERECALIBRATOR.out.table)
                .map { meta, bam, bai, table ->
                    [
                        meta,
                        bam,
                        bai,
                        table,
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.filtered_intervals.interval_list', checkIfExists: true)
                    ]
                }
                input[2] = Channel.value(
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                )
                input[3] = Channel.value(
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                )
                input[4] = Channel.value(
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        }
                        if(key == "bam") {
                            return [key, val.collect { [it[0], bam(it[1]).readsMD5, file(it[2].toString()).name]}]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }
}
