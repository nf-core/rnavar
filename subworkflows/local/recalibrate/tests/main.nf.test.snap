{
    "bam - no stats": {
        "content": [
            {
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "22e22cd4b7da6967236f1c83277ea7e3",
                        "test.recal.bam.bai"
                    ]
                ],
                "qc": [
                    
                ],
                "versions": [
                    "versions.yml:md5,5ac2b80142e8e414d486250387f61a89",
                    "versions.yml:md5,7254d01c09ca362f258ee3dda431ac92"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-08T10:20:14.080519845"
    },
    "bam - stats": {
        "content": [
            {
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "22e22cd4b7da6967236f1c83277ea7e3",
                        "test.recal.bam.bai"
                    ]
                ],
                "qc": [
                    [
                        {
                            "id": "test"
                        },
                        "test.stats:md5,eafd72617feeeb8644a72e4799281a09"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,40d9d05926d5b2862a346cf91e38655e",
                    "versions.yml:md5,5ac2b80142e8e414d486250387f61a89",
                    "versions.yml:md5,7254d01c09ca362f258ee3dda431ac92"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-08T10:19:51.240978907"
    }
}